### ISSUE TITLE
What about Map support as a functor?

### ISSUE THREAD

**Issue by ArturAralin** [ramda/ramda#2779]
```javascript
const srcMap = new Map([
  ['k1', 'v1'],
  ['k2', 'v'],
]);
```

```
const newMap = map(toUpper, srcMap);
newMap.get('k1') // "V1"
newMap === m // false
```

```
mapObjIndexed((v, k) => [k, v], srcMap);
// [['k1', 'v1'], ['k2', 'v']]
```

**Comment by CrossEye** [comment]
I would like to support `Map` and `Set` as much as we can post v1.0.  `Map` is clearly a functor.  I'm pretty sure `Set` is as well, although the fact that mapping could shrink the size in the latter seems strange.  (`map(square, new Set([-2, -1, 0, 1, 2])) //=> new Set([4, 1, 0])`)

This is on the Roadmap strawman document I'm slowly writing.

**Comment by ArturAralin** [comment]
> the fact that mapping could shrink the size in the latter seems strange

I think in the case of Set it acceptable, because this behavior caused by Set nature

**Comment by CrossEye** [comment]
My concern about `Set` is that a common description I use is that a Functor's map transforms a container of one type to a container of the same shape holding another type.  It makes sense, but "same shape" is not at all precisely defined.  `map` on arrays turns it into another array of the same length.  `map` on a Maybe turns it into the same sort of Maybe (Just/Nothing) as the original.  `map` can be defined on a tree, and the resulting tree has an identical branching structure.  It feels like that's the right way to think about it.

But it seems as though the most shape-like property of `Sets` is the count of elements.  And that's not preserved by `map`.  This is not an actual problem with Set or a related `map` if the laws are upheld.  But it is an unfortunate exception to the "same shape" heuristic.

The other thing to note is that what I really want to promote is the internal `Set` implementation and a comparable `Map`.  The JS global objects are probably not a good fit for Ramda, working with reference rather than value equality.

**Comment by ArturAralin** [comment]
I'm still not strong in theoretical basis of FP therefore hope that you make good decision. But second part of your respond causing the next question: 
Internal implementation does not conflict with the ramada philosophy?

**Comment by CrossEye** [comment]
No, it doesn't.  That's why we have the internal implementation: to be able to use something like `Set` which accepts Ramda's notion of equality:

```js
const set = new Set();
set.add({x: 1})
set.add({x: 1})
set.size //=> 2

const _set = new _Set();
_set.add({x: 1})
_set.add({x: 1})
_set.size //=> 1
```

This doesn't actually work right now in our internal implementation, but it would be relatively easy to support `size` by adding something like this:

```js
Object.defineProperties(_Set.prototype, {
  size: {
    get: function() { return Object.keys(this._items).length},
  }
})
```

We could also add support for the iteration protocols in the same way.

**Comment by dmitriz** [comment]
I think you can implement the Set Functor for any abstract equality relation, which basically means to extend the native JS equality used in `JS Set`, where the first example is correct since `{x:1} === {x:1}` evaluates to `false`, while the second relies on another equality. Imo both can be useful to support.

**Comment by CrossEye** [comment]
@dmitriz: While it would certainly be possible to support both, as a matter of prioritization it would make sense for Ramda to work with types that match its basic philosophy, and that includes an emphasis on value equality.

I'd hate to have to start answering the inevitable questions such as "Why does my Set not include the value I just added?" with discussions about `Set` vs `R.Set`.  Uggh!

**Comment by dmitriz** [comment]
> While it would certainly be possible to support both, as a matter of prioritization it would make sense for Ramda to work with types that match its basic philosophy, and that includes an emphasis on value equality.

Are you referring to the philosophy outlined here?
https://github.com/ramda/ramda#philosophy
It doesn't seem to mention anything about equality.
Or should I look elsewhere?

**Comment by CrossEye** [comment]
@dmitriz: I meant it in the general sense.  I wrote an article more than four years ago on [The Philosophy of Ramda][pr], but I don't think I spent much time even there on this specific question.

But the notion of value equality suffuses functional programming, and it's basic to Ramda.  It's why we had to create our own internal `Set` implementation.  It's why a few functions seem to be much less performant than their counterparts in other libraries.  (It's much easier to perform a `uniq` test when you only need to test reference equality!)

Ramda is always trying to strike a difficult balance between promoting good FP practices and working with the grain of Javascript.  In some places we move away from pure FP to make it more approachable for beginners.  But there are some things where we allow few compromises.  Along with immutability and referential transparency, value equality is one of them.


  [pr]: https://fr.umio.us/the-philosophy-of-ramda/

**Comment by dmitriz** [comment]
@CrossEye
I have read your blog long while ago and really liked how it made clear some of ramda's advantages and it was what attracted my interest to it at that time. But there is nothing there I could see on the specific issue of equality as you point out yourself. I would say, your article addressed points that are way more fundamental. Equality in comparison feels like a minor thing - a choice of equivalence relation and I have never used ramda's functions based on that. I can see from the source code that you have done quite a rework of the JS own (arguably imperfect) equality and it surely makes sense. But I could not find any documentation defining those changes.

> But the notion of value equality suffuses functional programming, and it's basic to Ramda.

Indeed, in FP you write generic code where you can plug any functor or monad, then why would you not regard equality relations as another similarly pluggable piece of modular structure? In fact, this is how I understood ramda would work, precisely, because that was what I've thought was ramda's philosophy that I really liked. :)

> It's why we had to create our own internal Set implementation.

Are you referring to this code:
https://github.com/ramda/ramda/blob/v0.26.1/source/internal/_Set.js
Not sure I understand its relation with `equals` and at what extent should it be regarded as something defining ramda, being just an "internal" thing. Some lines feel puzzling e.g. does this make `-0` and `'-0'` regarded as equal?
https://github.com/ramda/ramda/blob/v0.26.1/source/internal/_Set.js#L44

**Comment by CrossEye** [comment]
@dmitriz:

Yes, that is our Set implementation, and it's used at least in `uniq` and `difference`.  One of the things it does it distinguish `0` from `-0`:

```js
uniq([0, -0]) //=> [0, -0]
``` 

I think the first easily digestible explanation of one of Ramda's goals was in https://github.com/ramda/ramda/issues/1067#issuecomment-135990250: 

> `x` ≍ `y` should imply `f(x)` ≍ `f(y)`

where  "`a` ≍ `b`" means the same as "`equals(a, b)`".

But it's something we've been striving for all along.  (We are not there for reasons partially explained in that long thread.)  It is something that motivates a lot of our design.  But if we start supporting `Set` and `Map` with their reference equality, we run into problems:

```js
const a = {x: 1}, b = {x: 1}, c = {x: 1}
const count = xs => new Set(xs).size

R.equals([a, a, a], [a, b, c])               //=> true
R.equals(count([a, a, a]), count([a, b, c])) //=> false (1 !== 3)
```

While clearly one can make a Functor out of `Map`  or `Set` with their reference equality, Ramda has to decide for what types it offers Functor's `map` operation.  We've chosen so far to implement it for Arrays, Objects, and Functions -- three of the types we support with many functions -- and to delegate it for types that declare their own `map` method.  The proposal here is to add `Map`.  The question then would be how else we want to support `Map`?  As I said above, I'm mostly in favor of doing this, but I do have very real concerns.

**Comment by dmitriz** [comment]
>  One of the things it does it distinguish 0 from -0.

Indeed, this makes more sense than the weird JS native behaviour, but is it documented anywhere?

> `x` ≍ `y` should imply `f(x)` ≍ `f(y)`
> where "a ≍ b" means the same as "equals(a, b)".

This obviously depends on how you define `equals` and on what kind of `f` is allowed. So if two objects are considered equal irrespective of the key order, then `Object.keys` obviously breaks that rule. But that is a JS native function you have no controll over. Otoh, you can ensure it whenever `f` is any ramda function, and if it is already ensured, perhaps it needs to be documented somewhere "officially" (not in some issue), so we can reliably refer to that place as fundamental assumption already agreed upon that we don't need to revisit.

On the other hand, there is another very natural and useful equivalence relation, where objects are only equivalent when they keys have the same order. That is, the order is considered as part of the structure. Which is in fact natively supported by JS. And that relation will make several important JS functions referential transparent that the currently used `equals` doesn't. Also it will allow to define `reduce` over objects, which would be also very useful and was requested as far as I know.

So we arrive at two different equivalences that are but both useful. That indicates that maybe we should not limit ourselves to one particular equivalence, when we are already using at least two (ramdas and JS native).

Addressing this should also help with the issue you mention:

```js
const count = xs => new Set(xs).size
```

This function is using JS native `Set` that is referentially transparent but for a different equality, not the one used internally by ramda. And as it breaks ramda's equality, why would you even expect the derived function `count` to preserve it? But you can still preserve it with ramda's native functions, but `Set` isn't one, so I don't quite understand how is this a problem.

And even some of ramda's own functions seem to break it:
https://ramdajs.com/repl/#?keys%28%7Ba%3A1%2Cb%3A1%7D%29%20%3D%3D%3D%20keys%28%7Bb%3A1%2Ca%3A1%7D%29%20%2F%2F%3D%3E%20false%0A
```js
keys({a:1,b:1}) === keys({b:1,a:1}) //=> false
```

On the other hand, this ref. transparency would be restored, had `keys` returned a `Set` instead of an array. 

> The proposal here is to add Map. The question then would be how else we want to support Map? As I said above, I'm mostly in favor of doing this, but I do have very real concerns.

As Profunctor, it should perhaps support `map` and `comap` (or `contramap`) delegating to the more general `promap`.  As equivalent list of pairs, it could support `mapKeyed`.  As Foldable, it could support`reduce` and `reduceKeyed`. And there should be less problems with equality, the key order being part of the structure.

> And while the idea from your initial comment, supporting multiple Functors on sets with different equality relations is interesting, I'm not sure what that would mean for Ramda. 

Adding `equality` as optional parameter? Including the JS native option? And defaulting to the ramda's chosen equality. I'd find it helpful as it would instantly alert the user to the different equality that appears somewhat "hidden" right now. Alternatively you could duplicate with functions like `uniqEq` if you have optional parameters :)

> Obviously, the native Set really only supports reference equality. We could have a type `InternalSet` parameterized by an equality operation, but I don't see much of a useful API from that. Do you?

Maybe something like `SetEq` parametrized by the equality? Supporting the set functor/monad could be useful e.g. for any nondeterministic computation, even for operations as basic as `Object.keys`.

**Comment by dmitriz** [comment]
Another useful notion of equality that is not even an equivalence relation would be `eqApprox` that would satisfy

```js
eqApprox(0.1+0.2, 0.3) //=> true
R.equals(0.1+0.2, 0.3) //=> false
0.1+0.2 == 0.3 //=> false
```

**Comment by CrossEye** [comment]
@dmitriz:

Can we back up a bit here and detail a little more what you think would be a good idea?  I'm afraid we might be talking past one another.

What would it mean for you for Ramda to support multiple equivalence relations?

----------

It certainly is a good point that 

> `x` ≍ `y` should imply `f(x)` ≍ `f(y)`

Can't be universal in JS unless `≍` is the native reference equality, and even then it could fail for impure functions.  If you happened to read all of that long thread I mentioned, you'll see that I suggested a fix (key-sorting) that would at least make it true for all Ramda functions, one that I thought would be widely accepted, but which turned out to be too controversial to add.  That would address the problem with `keys` and certain other functions.  I'm still in favor of it, and when I spend more time on Ramda, I will try to argue for it again.

----------

I'm not opposed to adding `Set` and `Map` support.  I thought we would have done that long ago.  But every time I try to think about doing it, I run into all sorts of complications.  Just adding them to `equals` are problematic enough, but trying to deal with them as Functors and the like is scarier still.

But perhaps your idea of what it would mean to support reference equality might clear something up for me.
